pkg/ebpf/dns_snooper.go

```diff
@@ -0,0 +1,206 @@
+)
+
+const (
+	dnsCacheTTL              = 3 * time.Minute
+	dnsCacheExpirationPeriod = 1 * time.Minute
+	dnsCacheSize             = 100000
+	packetBufferSize         = 100
+)
+
+var _ ReverseDNS = &SocketFilterSnooper{}
+
+// SocketFilterSnooper is a DNS traffic snooper built on top of an eBPF SOCKET_FILTER
+type SocketFilterSnooper struct {
+	tpacket      *afpacket.TPacket
+	source       *gopacket.PacketSource
+	socketFilter *bpflib.SocketFilter
+	socketFD     int
+	ipsToNames   *reverseDNSCache
+	exit         chan struct{}
+	wg           sync.WaitGroup
+}
+
+// NewSocketFilterSnooper returns a new SocketFilterSnooper
+func NewSocketFilterSnooper(filter *bpflib.SocketFilter) (*SocketFilterSnooper, error) {
+	tpacket, err := afpacket.NewTPacket(afpacket.OptPollTimeout(1 * time.Second))
+	if err != nil {
+		return nil, fmt.Errorf("error creating raw socket: %s", err)
+	}
+
+	packetSrc := gopacket.NewPacketSource(tpacket, layers.LayerTypeEthernet)
+
+	// Unfortunately the underlying socket file descriptor is private
+	socketFD := int(reflect.ValueOf(tpacket).Elem().FieldByName("fd").Int())
+
+	err = bpflib.AttachSocketFilter(filter, socketFD)
+	if err != nil {
+		return nil, fmt.Errorf("error attaching filter to socket: %s", err)
+	}
+
+	reverseDNSCache := newReverseDNSCache(dnsCacheSize, dnsCacheTTL, dnsCacheExpirationPeriod)
+	snooper := &SocketFilterSnooper{
+		tpacket:      tpacket,
+		source:       packetSrc,
+		socketFilter: filter,
+		socketFD:     socketFD,
+		exit:         make(chan struct{}),
+		ipsToNames:   reverseDNSCache,
+	}
+
+	// Start consuming packets
+	packetChan := snooper.createPacketStream(1000)
+	snooper.wg.Add(1)
+	go func() {
+		snooper.run(packetChan)
+		snooper.wg.Done()
+	}()
+
+	return snooper, nil
+}
+
+// Resolve IPs to Names
+func (s *SocketFilterSnooper) Resolve(connections []ConnectionStats) []NamePair {
+	return s.ipsToNames.Get(connections, time.Now())
+}
+
+// Close terminates the DNS traffic snooper as well as the underlying socket and the attached filter
+func (s *SocketFilterSnooper) Close() {
+	close(s.exit)
+	s.wg.Wait()
+	err := bpflib.DetachSocketFilter(s.socketFilter, s.socketFD)
+	if err != nil {
+		log.Errorf("error detaching socket filter: %s", err)
+	}
+	s.tpacket.Close()
+	s.ipsToNames.Close()
+}
+
+func (s *SocketFilterSnooper) run(packets <-chan gopacket.Packet) {
+	for packet := range packets {
+		layer := packet.Layer(layers.LayerTypeDNS)
+		if layer.LayerType() != layers.LayerTypeDNS {
+			continue
+		}
+		dns, ok := layer.(*layers.DNS)
+		if !ok {
+			continue
+		}
+
+		translation := parseAnswer(dns)
+		if translation == nil {
+			continue
+		}
+
+		s.ipsToNames.Add(translation, time.Now())
+	}
+}
+
+func (s *SocketFilterSnooper) createPacketStream(chanSize int) <-chan gopacket.Packet {
+	packetChan := make(chan gopacket.Packet, packetBufferSize)
+	go func() {
+		defer close(packetChan)
+		for {
+			packet, err := s.source.NextPacket()
+			if err == nil {
+				packetChan <- packet
```

Comment:
Resolving this. I validated that we're protected from the scenario you mentioned since the sender go routine in this case is also responsible for closing the channel.